### PR TITLE
feat: Tensorboard use local storage [DET-6029]

### PR DESCRIPTION
**Pull request by varlogtim**
Features:
 - Tensorboard rewrite to sync to local storage.
 - Start of named logger use
 - Added debug flag for task specs. Only used by Tensorboard.
 - Tensorboard now starts for any new files found, i.e., profiler fix

Tensorboard Fetchers:
 - GCS Fetcher
 - S3 Fetcher
 - Azure Fetcher
 - SharedFS

Bug Fixes:
 - Empty directories in checkpoints did not work for Azure
 - Upload for azure_client should overwrite to match S3 behavior.

Unit Tests added:
 - Azure live storage manager test.
 - S3 Tensorboard Fetcher
 - GCS Tensorboard Fetcher
 - Azure Tensorboard Fetcher
 - SharedFS Tensorboard Fetcher

### CODE REVIEW

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [inline harness/determined/exec/tensorboard.py:83]: Why wait for tensorboard to show metrics, when we now know perfectly well when metrics are ready?

File: harness/determined/tensorboard/fetchers/s3.py
Comment (rb-determined-ai) [inline harness/determined/tensorboard/fetchers/s3.py:47]: Is this going to become very expensive on large experiments?

I wonder if we could save tensorflow files with timestamps in the filename and ignore anything that was too old using the `start_from` parameter or whatever it's called.

Just a though, not in scope for this PR I think.

File: harness/tests/storage/util.py
Comment (rb-determined-ai) [inline harness/tests/storage/util.py:159]: This is a bit of a strange cleanup pattern, and it's not clear that it's bulletproof.

Why not just create a temporary directory, and put all the files you are going to clean up in there, then delete the whole thing in the `finally` block?

**Review by rb-determined-ai — COMMENTED** [review]
Looks pretty good.  I like the tests a lot.

Regarding your request for having a StorageInterface: I still think that the "right" abstraction are the ones we already have.  You want a Fetcher?  We have a Fetcher interface.  You want a StorageManager?  We have a storage manager.  The interface you outlined to me in DM looks tailor-made for blob storage, but it looks like the shared-fs storage implementations would get worse, not better.

What would you say to leaving the existing interfaces (Fetcher, StorageManager), but implementing Azure/GCS/S3 as a BlobFetcher or BlobStorageManager?  Each of those could take a BlobBackend, with methods as you describe in your StorageInterface, but since the new interface would be specific to blob storage, I think you'd end up with a lot cleaner code.

File: master/pkg/model/storage_config.go
Comment (varlogtim) [reply master/pkg/model/storage_config.go:1]: I actually don't know where this came from at all. I must have written it at some point to add master config validation for storage related components during master startup.

Comment (rb-determined-ai) [reply master/pkg/model/storage_config.go:1]: probably a rebasing issue... I've deleted this file since the very first pr you posted a long while ago.

File: harness/tests/storage/util.py
Comment (varlogtim) [reply harness/tests/storage/util.py:143]: done!

File: harness/determined/common/storage/azure.py
Comment (varlogtim) [reply harness/determined/common/storage/azure.py:67]: I don't recall doing so. Must have been an accident.

File: harness/determined/common/storage/azure_client.py
Comment (varlogtim) [reply harness/determined/common/storage/azure_client.py:31]: Yeah, I teetered on a few different options here. Updated to only specify the requirements, which is the condition.

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:83]: I am not sure what you mean here, could you add a little clarity?

You could me that we know if we have downloaded files or not, so why wait to see if Tensorboard is responding with metrics?

The short answer is, this was how it was working.

File: harness/determined/tensorboard/fetchers/s3.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/s3.py:47]: > Is this going to become very expensive on large experiments?
I am not sure what you mean by expensive. Do you mean for situation where there are ~10K files, we will do 10 requests per second, to list files?

So, you are thinking that at some point after the initial listing of files we then start giving a time offset to the `list_objects_v2()` call so that we only see recently updated files? This is a super good idea actually!

I am curious if Tensorboard does this already.

> not in scope for this PR I think
Cool. Will do it at some point in the future.

File: harness/tests/storage/util.py
Comment (varlogtim) [reply harness/tests/storage/util.py:159]: > This is a bit of a strange cleanup pattern, and it's not clear that it's bulletproof.

This will cause the test to fail if we fail to remove one of the files. Thoughts on this behavior?

I basically need to delete a dynamic list of files when I exit the test context for any reason. I thought about using a context manager(s) here, but it felt strange because the list is updating and several files get replaced, so don't need to be removed. 

(I probably should have used a dict instead of a list though in the event that the save file is appended to the list twice)


> Why not just create a temporary directory, and put all the files you are going to clean up in there, then delete the whole thing in the finally block?

These are files that are present inside an unknown storage backend. I needed to pass in the storage manager specific `put()` and `rm()` so that I didn't have to copy/paste the test for-each storage backend we have and change 2 lines, like the Fetchers.

**Comment by varlogtim** [comment]
To be clear, I think this is super unimportant and doesn't matter much except in certain situations. You can completely ignore this comment.

> Regarding your request for having a StorageInterface: I still think that the "right" abstraction are the ones we already have. You want a Fetcher? We have a Fetcher interface. You want a StorageManager? We have a storage manager.

I may have used the wrong words the first time to explain what I was thinking so I wanted to give it a second shot. (Or maybe you understand my position but I still don't understand fully understand yours and instead of asking for clarity, I am restating my own position... ewww... If this is the case, sorry.)

When we interact with persistent storage we do so as if it was a hierarchical file system, which I will refer to as "HFS abstraction." I am proposing explicitly defining this HFS abstraction. Due to the current structure, we are forced to re-implement the HFS abstraction both when we go to add a new storage back end and when we add new functionality which uses our storage back ends. In the case of adding a new storage back end, we are required to implement specially named `*StorageManager`, `*TensorbordManager` and `*Fetcher` classes, in which the only thing that is different from their respective counterparts, is the HFS abstraction components. When new functionality is added, we must copy and paste the new functionality code for each of the storage back ends, and again just change the HFS abstractions. What we end up with is the exact same HFS abstaction code existing in multiple different function specific places, and the exact same *Functionality code existing in multiple places.

I am not proposing to remove *StorageManager, *TensorboardManager, and/or *Fetcher abstractions; I am just proposing a separation of the HFS abstraction and the functionality of each *Function. This will allow these *Functions, as well as our tests, and any future *Fuctions, can more cleanly and clearly interact along the HFS assumptions that these *Functions are already making internally.

> The interface you outlined to me in DM looks tailor-made for blob storage, but it looks like the shared-fs storage implementations would get worse, not better.

It isn't tailor-made for blob storage, it is made to make any persistent storage look like a hierarchical filesystem and we happen to only have implementations for blob storage at the moment. My intention is to make interacting with any persistent storage, database, blob storage, REST service, etc.. as easy as it is with SharedFS. It will be much easier for us, or users, to add storage back ends, without the need to implement a, potentially ever growing, set of the *Function classes. It also paves the way for an endpoint like, `add_storage_backend(name, code)`, which would create a storage backend without bringing down the cluster and rebuilding.
 
> What would you say to leaving the existing interfaces (Fetcher, StorageManager), but implementing Azure/GCS/S3 as a BlobFetcher or BlobStorageManager? Each of those could take a BlobBackend, with methods as you describe in your StorageInterface, but since the new interface would be specific to blob storage, I think you'd end up with a lot cleaner code.

The seems like a good idea and there are likely Blob specific components to this.

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [reply harness/determined/exec/tensorboard.py:82]: > You could me that we know if we have downloaded files or not, so why wait to see if Tensorboard is responding with metrics?

Yes.  I know how it was working but there's a better way now, right?

File: harness/determined/tensorboard/fetchers/azure.py
Comment (rb-determined-ai) [reply harness/determined/tensorboard/fetchers/azure.py:70]: wait... no `+` doesn't indicate that you are appending, it indicates that you are reading and writing.  Still seems odd here.

File: harness/determined/tensorboard/fetchers/s3.py
Comment (rb-determined-ai) [reply harness/determined/tensorboard/fetchers/s3.py:46]: > Do you mean for situation where there are ~10K files, we will do 10 requests per second, to list files?

Yes.  If each file were 50 bytes long (a low estimate given that there's a uuid in the mix) then that would be 1MB/s (8Mbps), which is probably on the order of streaming an hd movie, just to sit there doing nothing.  It's clearly an extreme example, but the point is that this strategy does not scale well.

File: harness/tests/storage/util.py
Comment (rb-determined-ai) [reply harness/tests/storage/util.py:159]: > I basically need to remove a dynamic list of files from an unknown storage when I exit the test context for any reason

Not if you run the test in a temporary directory you create at the top of the test.  Then you just need to `shutil.rmtree()` a single directory.

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:82]: What would be the better way? Just launch it? Say the word and I will. It solves another issue which is that I think it hangs when someone is trying to use the profiler.

File: harness/determined/tensorboard/fetchers/s3.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/s3.py:46]: Do you want me to change and test this before this goes in? Or can I commit to do it after this lands? I might need to change the tests.

File: harness/tests/storage/util.py
Comment (varlogtim) [reply harness/tests/storage/util.py:159]: I don't think you are understanding that these `files_to_remove` exist in S3, Azure or GCS. Would you rather me not test against S3, Azure, and GCS? Or maybe I am not understanding what you mean?

File: harness/determined/common/storage/azure.py
Comment (varlogtim) [reply harness/determined/common/storage/azure.py:67]: Fixed.

File: master/pkg/model/storage_config.go
Comment (varlogtim) [reply master/pkg/model/storage_config.go]: Removed the file. Even though I am not sure why this file isn't in the source already.

File: harness/determined/tensorboard/fetchers/s3.py
Comment (rb-determined-ai) [reply harness/determined/tensorboard/fetchers/s3.py:46]: Um, I'm actually not 100% certain how you would fix it.  Maybe backoff on the frequency of updates as they start to grow very large?

Anyway, no I think it's probably ok to land as-is.  But I don't feel totally confident... perhaps get another opinion, from @stoksc or @ioga?

File: harness/determined/tensorboard/fetchers/s3.py
Comment (ioga) [inline harness/determined/tensorboard/fetchers/s3.py:40]: could you please use boto paginator machinery instead of doing it manually: https://boto3.amazonaws.com/v1/documentation/api/latest/guide/paginators.html

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:122]: Talked about this and will write to temp.

File: harness/determined/tensorboard/fetchers/__init__.py
Comment (rb-determined-ai) [inline harness/determined/tensorboard/fetchers/__init__.py:29]: not sure what this means, but please resolve it

File: harness/determined/tensorboard/fetchers/s3.py
Comment (ioga) [reply harness/determined/tensorboard/fetchers/s3.py:46]: it's hard to judge without knowing the actual usage patterns. 
- ship it today but add more logging to catch the "expensive" cases? i.e. can we add `info` level logging for "large" responses here so we could catch it when it happens?
- strategy-wise - maybe we could backoff and poll more rarely when there're no changes to reduce the idle traffic?
- would more advanced solutions around s3 events -> SNS/SQS be viable?

File: harness/tests/storage/util.py
Comment (rb-determined-ai) [reply harness/tests/storage/util.py:159]: talked on zoom, `rm_files` is about storage backend not local files, and this seems reasonable to me

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:82]: So, as soon as one download happens, I will trigger the ready message.

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [inline harness/determined/exec/tensorboard.py:104]: seems like the TRIGGER_READY_MSG is probably sufficient already

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [inline harness/determined/exec/tensorboard.py:90]: Why not combine these strings?  Or can you put a `+` in between the strings?

File: harness/determined/tensorboard/fetchers/s3.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/s3.py:46]: I am not sure how/where we would get the info logs for this?

I will commit to looking into investigating if:
1. Tensorboard files are only written to once and not updated (this isn't clear to me right now).
2. Dig into source for `tensorboard/backend/event_processing/data_ingester.py` and see what they do there.
3. Assuming only new files are ever written, that we rename the files so that they are sortable for use with the `StartAfter` flag.
4. Look for corresponding options in the other storage platforms we support.

Also, @rb-determined-ai, this is another example where a generic storage interface would suffer in performance :)

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:90]: This was an annoying MyPy formatting thing. It doesn't like it when I printf a string and bytes, even though I intent to print bytes. Are you cool with that?

Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:90]: Done. This was a weird MyPy issue along with auto-formatting strangeness.

File: harness/determined/tensorboard/fetchers/s3.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/s3.py:40]: Oh nice! I don't think this existed the last time I wrote code for boto3.

File: harness/determined/tensorboard/fetchers/__init__.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/__init__.py:29]: I think I resolved but didn't remove message.

File: harness/determined/common/_logging.py
Comment (varlogtim) [reply harness/determined/common/_logging.py]: done.

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py]: done.

File: harness/determined/tensorboard/fetchers/azure.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/azure.py]: Resolved.

File: harness/determined/tensorboard/fetchers/s3.py
Comment (rb-determined-ai) [reply harness/determined/tensorboard/fetchers/s3.py]: I'd say:

2. Good idea, but IMO only useful if you find out that they're doing something better than us.  If what they do is as naive as us, then that isn't really good enough, since we are embarking on this mission because we're sick of dealing with tensorboard not behaving very well in corner case (that is, we really need to be better)
3. Not a good idea.  Tensorboard fetchres should be read-only.

That SNS/SQS that Ilia mentioned would be a sweet addition.  I spent some time and arrived poking around [this](https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html#supported-notification-event-types) documentation, and I don't think that would really be meaningful for us to support any time soon.

Also, @varlogtim, why did you mark this comment thread as "resolved" when it is clearly still active?

> I am not sure how/where we would get the info logs for this?

No, he means you should be logging fetches which become very large at the info level, so it's more obvious if we have a problem.   Maybe just log any time we do >1 fetch per second?

File: master/pkg/model/storage_config.go
Comment (rb-determined-ai) [reply master/pkg/model/storage_config.go]: I removed it, when I removed the old ExperimentConfig object that was hardly in use.  The new one in `pkg/schemas/expconf` has several better properties.

File: harness/determined/common/_logging.py
Comment (rb-determined-ai) [reply harness/determined/common/_logging.py]: you say "done" and you marked this as resolved, but you didn't actually handle my biggest complaint, which is that:

> It's configuring formatters and stdout handlers, neither of which should be configured by the library

File: harness/determined/tensorboard/fetchers/azure.py
Comment (rb-determined-ai) [inline harness/determined/tensorboard/fetchers/azure.py:19]: remove all of these `, None`s

File: harness/determined/tensorboard/fetchers/azure.py
Comment (rb-determined-ai) [inline harness/determined/tensorboard/fetchers/azure.py:21]: ditto.  I'm going to stop marking them, please find all of them.

File: harness/determined/tensorboard/fetchers/azure.py
Comment (rb-determined-ai) [inline harness/determined/tensorboard/fetchers/azure.py:56]: most of the `logger.debug()` statements in these `fetchers/*` files strike me as "useful while developing the feature but would probably just be noise if I turned logging to debug in real life".

You do more support than I do, so if you feel these should stay I won't press the issue.  But I want you to consider them carefully, because unhelpful logging is not without cost.  This one in particular seems like it would be in a pretty hot loop, but would be particularly useless.

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [inline harness/determined/exec/tensorboard.py:123]: Why even catch the error?  When there is an error you print the error text (but not the stack trace), then you kill the tensorboard and exit with the exit code of the tensorboard process.  If you didn't catch the error, you would instead exit printing an entire stack trace, which seems preferable to me.

I'd much rather see a whole stacktrace than a `NoneType error` or some unhelpful string that could have originated in a variety of different files.

Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:123]: I think it is left over code, which doesn't make sense, from the review process. Removed.

File: harness/determined/tensorboard/fetchers/azure.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/azure.py:56]: I agree. Removed.

File: harness/determined/tensorboard/fetchers/azure.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/azure.py:21]: Yup, I suppose this is a bad habit.

File: harness/determined/tensorboard/fetchers/__init__.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/__init__.py:30]: :+1:

File: harness/determined/tensorboard/fetchers/__init__.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/__init__.py:27]: done.

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:13]: Oh, nice!

File: harness/determined/common/_logging.py
Comment (varlogtim) [reply harness/determined/common/_logging.py]: I think your comments more after we discussed that you don't actually want people to know which module/file is doing this logging. I will just use the existing `set_logger()`.

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [inline harness/determined/exec/tensorboard.py:134]: you probably need `if not responsive: ...` here.  No point in checking tensorboard over and over I think.

File: harness/determined/exec/tensorboard.py
Comment (rb-determined-ai) [inline harness/determined/exec/tensorboard.py:127]: Your if statement and the message you give to the user do not match.

Probably just update the message?

File: harness/determined/tensorboard/fetchers/shared.py
Comment (rb-determined-ai) [inline harness/determined/tensorboard/fetchers/shared.py:13]: Just had a thought: Why even have a SharedFS Fetcher?  Why not just use the files right in the shared_fs?

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:134]: Ahh, yeah, I reworked this and that slipped through... it was there before somewhere.

File: harness/determined/tensorboard/fetchers/shared.py
Comment (varlogtim) [reply harness/determined/tensorboard/fetchers/shared.py:13]: Good-ish reasons:
Consistency.
I can't think of a good reason to directly access `shared_fs`.
I do think that there will be an odd exception in the tensorboard code for shared_fs in which it needs to worry about the storage type.

Not-so-good-reasons:
I would like this to go in by Wednesday. Can we make that change at a later point?

If required for your approval, please tell me and I will make the change.

File: harness/determined/exec/tensorboard.py
Comment (varlogtim) [reply harness/determined/exec/tensorboard.py:127]: Yup.

File: harness/determined/tensorboard/fetchers/shared.py
Comment (rb-determined-ai) [reply harness/determined/tensorboard/fetchers/shared.py:13]: > I can't think of a good reason to directly access shared_fs.
Less code, and less disk io?

> Can we make that change at a later point?
Yes, I think that's fine.

**Review by rb-determined-ai — APPROVED** [review]
great work!  especially thank you for writing tests!

**Pull request merged**
